sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample13_AuthenticateWithClientSecretCredential.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+using Azure.Identity;
+using Azure.Messaging.EventHubs.Samples.Infrastructure;
+
+namespace Azure.Messaging.EventHubs.Samples
+{
+    public class Sample13_AuthenticateWithClientSecretCredential : IEventHubsIdentitySample
+    {
+        /// <summary>
+        ///   The name of the sample.
+        /// </summary>
+        ///
+        public string Name { get; } = nameof(Sample13_AuthenticateWithClientSecretCredential);
+
+        /// <summary>
+        ///   A short description of the sample.
+        /// </summary>
+        ///
+        public string Description { get; } = "An example of authentication using ClientSecretCredential.";
+
+        /// <summary>
+        ///   Runs the sample using the specified Event Hubs connection information.
+        /// </summary>
+        ///
+        /// <param name="fullyQualifiedNamespace">The fully qualified Event Hubs namespace.  This is likely to be similar to <c>{yournamespace}.servicebus.windows.net</param>
+        /// <param name="eventHubName">The name of the Event Hub, sometimes known as its path, that the sample should run against</param>
+        /// <param name="tenantId">The Azure Active Directory tenant that holds the service principal</param>
+        /// <param name="clientId">The Azure Active Directory client identifier of the service principal</param>
+        /// <param name="secret">The Azure Active Directory secret of the service principal</param>
+        ///
+        public async Task RunAsync(string fullyQualifiedNamespace,
+                                   string eventHubName,
+                                   string tenantId,
+                                   string clientId,
+                                   string secret)
+        {
+            // Service principal authentication is a mean for applications to authenticate
+            // against Azure Active Directory (AAD) and consume Azure services. This is advantageous compared
+            // to signing in using fully privileged users as it allows to enforce role-based authorization
+            // from the portal.
```

Comment:
It also allows for rolling of credentials without tearing down clients that may be running - something that is difficult if not impossible for connection strings.

Reply:
I am not gonna lie to you, I think I lost the plot here. I saw this [issue](https://github.com/Azure/azure-sdk-for-js/issues/1732) does this mean that if I am changing an SAS key in a connection string AMQP connections with all the clients will be disrupted while if I change the secret connected to a service principal all the clients will be able to communicate until their token won't expire?

Reply:
There's a couple of things:

- Rolling credentials for a token-based user is something that was designed to be supported without disrupting an application.   As a client library, we want to encourage folks to use the Identity-based paths for production applications in the name of best practices for security and also for stability of the application.

- You're correct in that rolling credentials for the shared key in the portal will cause existing connections to become unauthorized at some point shortly after it is done; I don't know the specifics, as that's a service implementation detail.  By rolling the credential, you're essentially telling the service to disallow anything using the old key.

- When you provide a connection string to the client library, you're giving it a hard-coded key.  If you roll that key and need to update, you have to destroy your active Event Hub client instances and instantiate new ones with the new key.

- The issue that you linked was the result of some thought exercises on how we could support rolling a shared key in place without requiring the full-on "build another client" flow.  That said, it is also one of the big reasons that we're working hard to encourage developers to use an identity-based constructor for production applications.

- We do have a credential type that allows a shared key to be used as an identity, separate of a connection string.  We've got an [item](https://app.zenhub.com/workspaces/azure-sdk-team-5bdca72c4b5806bc2bf0aab2/issues/azure/azure-sdk-for-net/6618) to offer support on that to provide a new shared key on-the-fly, but developers would still need to do some work to plumb that into their code, where the AAD identity rolling is more transparent.

Did that help at all or just further muddle the issue?

Reply:
Thank you very much for your answer that was amazing.

That task is very interesting.

Reply:
> * but developers would still need to do some work to plumb that into their code, where the AAD identity rolling is more transparent

To see if I got it right, do you think this line would work?

> Another huge advantage compared to using Connection Strings is that rolling of credentials
> is a feature of OAuth 2.0 and implemented by Azure Active Directory. Whenever a token expires or is revoked, the SDK will take care of issuing a new one transparently. The same would not be true with connection strings and changing credentials will require the EventHubProducerClient or EventHubProducerConsumer to be disposed and recreated.